javalanglib/src/main/scala/java/lang/ClassValue.scala

```diff
@@ -0,0 +1,101 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package java.lang
+
+import java.util.HashMap
+
+import scala.scalajs.js
+import scala.scalajs.runtime.linkingInfo
+
+abstract class ClassValue[T] protected () {
+  import ClassValue._
+
+  private val jsMap: AnyRef = {
+    if (linkingInfo.assumingES6 || js.typeOf(js.Dynamic.global.Map) != "undefined")
+      js.Dynamic.newInstance(js.Dynamic.global.Map)(): AnyRef
+    else
+      null
+  }
+
+  @inline
+  private def useJSMap: scala.Boolean = {
+    /* The linking-info test allows to constant-fold this method as `true` when
+     * emitting ES 2015 code, which allows to dead-code-eliminate the branches
+     * using `HashMap`s, and therefore `HashMap` itself.
+     */
+    linkingInfo.assumingES6 || jsMap != null
+  }
+
+  /* We use a HashMap instead of an IdentityHashMap because the latter is
+   * implemented in terms of the former anyway, to a HashMap is leaner and
+   * faster.
+   */
+  private val javaMap: HashMap[Class[_], Any] =
+    if (useJSMap) null
+    else new HashMap()
+
+  protected def computeValue(`type`: Class[_]): T
+
+  def get(`type`: Class[T]): T = {
+    if (useJSMap) {
+      val existing: Any = jsMap.asInstanceOf[js.Dynamic].get(
+          `type`.asInstanceOf[js.Any])
+      existing match {
+        case () =>
+          val newValue = computeValue(`type`)
+          val patchedValue = if (() == newValue) Sentinel else newValue
+          jsMap.asInstanceOf[js.Dynamic].set(
+              `type`.asInstanceOf[js.Any], patchedValue.asInstanceOf[js.Any])
+          newValue
+        case Sentinel =>
+          ().asInstanceOf[T]
+        case _ =>
+          existing.asInstanceOf[T]
+      }
+    } else {
+      javaMap.computeIfAbsent(`type`, new ComputeFun(this)) match {
+        case Sentinel => null.asInstanceOf[T]
+        case value    => value.asInstanceOf[T]
+      }
+    }
+  }
+
+  def remove(`type`: Class[_]): Unit = {
+    if (useJSMap)
+      jsMap.asInstanceOf[js.Dynamic].delete(`type`.asInstanceOf[js.Any])
+    else
+      javaMap.remove(`type`)
+  }
+}
+
+private object ClassValue {
+  /** Sentinel used instead of `()` (`undefined`) in a JS `Map`, or instead of
+   *  `null` in a Java `HashMap`.
+   */
+  private object Sentinel
```

Comment:
I don't understand why we need this. Don't both `ju.HashMap` and `js.Map` support checking for values? (I realize `computeIfAbsent` cannot be used for this, but IMO then we should just not use it).

Reply:
They do, but it seems better to do a simple `===` against a sentinel than to do a full second lookup.

Reply:
Better for speed? Maybe. But not better for codesize (IIUC) or readability, so I'm not 100% sure we're making the right tradeoff here :-/

Reply:
All right. I switched a different implementation that does not use a sentinel.